feat: chainflip swapper

gomesalexandre opened the pull request:
## Description

Authored to be able to see diff, WIP. 
Takes over the monster work done in https://github.com/shapeshift/web/pull/7991 by @CumpsD 🐐 

- [x] cleanup
- [x] if https://github.com/shapeshift/web/pull/7997 goes in first (which it will, most likely), split rates and quotes
- [x] status endpint 🍺 
- [x] double-check status endpoint is happy for UTXOs sends and receives, not only EVMs
- [x] make it work (ensure execution is happy)
  - [x] EVM native assets (send/receive)
  - [x] EVM tokens (send/reveive)
  - [ ] BTC (send/receive) 
    - [x] Receive
    - [ ] https://github.com/shapeshift/web/issues/8115 ~~Sends - currently blocked by Taproot work in https://github.com/shapeshift/web/issues/6761~~
  - [x] Solana
- [x] mixpanel - no need to, already working as part of the swapper abstraction

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes https://github.com/shapeshift/web/issues/8026

## Risk

> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

Test all Chainflip supported assets (or at the very least one native asset and one token per chain family), both on the buy and sell side - see https://swap.chainflip.io for reference on supported assets

- Bitcoin (BTC) - unsupported for sells for the time being, see checkboxes above
- Ethereum (ETH, FLIP, USDC, USDT)
- Solana (SOL, USDT)
- Arbitrum (ETH, USDC)

- Ensure swaps are happy e2e, including status detection
- Ensure funds land on the correct account
- Ensure no approval step is required for tokens (this isn't a contract call, but a regular send)

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations

- [x] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)

- BTC Receive

<img width="539" alt="image" src="https://github.com/user-attachments/assets/44218f6a-da13-4434-aa1d-c57cf4cb202a">
<img width="570" alt="image" src="https://github.com/user-attachments/assets/563bc76a-0676-437f-b1ed-ce4251836b7d">
<img width="577" alt="image" src="https://github.com/user-attachments/assets/a767eb9b-6f2e-4add-9603-acda09ebe29b">
<img width="573" alt="image" src="https://github.com/user-attachments/assets/f7c7edc0-11cf-4cd3-b465-798302951ebe">
<img width="562" alt="image" src="https://github.com/user-attachments/assets/0aef2da4-c5c7-440d-b94f-ebb0f80918d4">
<img width="459" alt="image" src="https://github.com/user-attachments/assets/11273b17-401d-4f81-9332-f82ae62656c2">

- BTC Sends

Blocked for the time being

- Native EVM asset send

https://jam.dev/c/bf0f392f-64c1-40ac-a0e6-7a96809e954d

- EVM token send

https://jam.dev/c/9735bf9d-61aa-404d-96eb-d7e2aa92c0cd

- Solana receives

![image](https://github.com/user-attachments/assets/14090a1e-4ca7-4698-a185-39205166271c)

- Solana sends

<img width="602" alt="image" src="https://github.com/user-attachments/assets/c773f910-33df-47d4-a257-877c6c0309d2">

<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":""}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/utils/helpers.ts:
```diff
@@ -0,0 +1,25 @@
+  )
+}
+
+export const getGasLimit = (chainflipAsset: string) => {
```
This won't generalize and should be programmatic. Since we're talking sends, allowance isn't a problem - there will never be allowance required (which, on a tangent, means we should be cognicent in ensuring allowance step is never triggered in swapper)

gomesalexandre commented on packages/caip/src/constants.ts:
```diff
@@ -17,16 +17,22 @@ export const arbitrumNovaAssetId: AssetId = 'eip155:42170/slip44:60'
-export const foxyAssetId: AssetId = 'eip155:1/erc20:0xdc49108ce5c57bc3408c3a5e95f3d864ec386ed3'
-
 export const usdtAssetId: AssetId = 'eip155:1/erc20:0xdac17f958d2ee523a2206206994597c13d831ec7'
+export const usdcAssetId: AssetId = 'eip155:1/erc20:0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'
+export const usdcOnArbitrumOneAssetId: AssetId =
+  'eip155:42161/erc20:0xaf88d065e77c8cc2239327c5edb3a432268e5831'
+export const usdcOnSolanaAssetId: AssetId =
+  'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'
+export const flipAssetId: AssetId = 'eip155:1/erc20:0x826180541412d574cf1336d22c0c0a287822678a'
```
all common AssetIds moved into one block + some additions

CumpsD commented on .env.base:
```diff
@@ -169,3 +171,7 @@ REACT_APP_ZERION_BASE_URL=https://api.proxy.shapeshift.com/api/v1/zerion
 REACT_APP_ZRX_BASE_URL=https://api.proxy.shapeshift.com/api/v1/zrx/
+
+# chainflip
+REACT_APP_CHAINFLIP_API_KEY=6ba154d4-e219-472a-9674-5fa5b1300ccf
```
Give me a ping to set up Shapeshift's API key when going forward, this is currently mine for testing! :) The best person to reach out to me for this would be the one who will be responsible for redeeming affiliate rewards in the future. The one controlling the wallet to sign Chainflip transactions that is (redeeming is a chainflip tx)

gomesalexandre replied:
Sounds good @CumpsD! Is there any reason we should wait for go-live to get an actual API key here, or can we already get one? Will get the convs going already

CumpsD replied:
The key in there right now is actually the one from my swapping Discord bot https://swappy.be/ :) It works, you can test with it, but it charges 0.15% fee (0.05 BaaS + 0.10 Swappy). I can make you a temp one that does only 0.05 (BaaS fee + 0% partner) if you want. You can go live with that as well and swap it out to the Shapeshift one afterwards. lmk

gomesalexandre replied:
@CumpsD just to be sure, we can parametrize fee bps right i.e this isn't tied to the API key used?

gomesalexandre replied:
Current one for dev is absolutely fine, so long as we go live on prod (and ideally on develop as a follow-up) with a proper API key 🙏🏽

CumpsD replied:
> @CumpsD just to be sure, we can parametrize fee bps right i.e this isn't tied to the API key used?

Yes, you can pass in `commissionBps` for both quote and opening a deposit channel (I think I added that in the PR already even! taking in affiliateBps from shapeshift)

CumpsD commented on packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,330 @@
+    )
+  }
+
+  const sellChainflipChainKey = `${sellAsset.symbol.toLowerCase()}.${
```
Another fyi, if you want to improve this. The assets are available here: https://chainflip-broker.io/assets as the `id` property in the format `asset`.`chain`, it just so happened all of them matched the `.symbol`, but perhaps you want to have this in a mapping in constants as well, like the chains?

CumpsD replied:
I added this to the axios cache by the way, if it helps :) Didn't use it though

gomesalexandre replied:
Think that's fine as a first pass IMO given there aren't many assets supported, but as Chainflip supports more assets, we'll definitely want to make things programmatic by hitting this endpoint so we don't have to manually add supported assets everytime!

CumpsD commented on packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,330 @@
+    const error = maybeQuoteResponse.unwrapErr()
+    const cause = error.cause as AxiosError<any, any>
+
+    if (
```
This can probably be done cleaner, the error responses follow the problem details rfc, maybe axios has some way to properly parse that to a typed thing instead of the `any` I had to do here

CumpsD replied:
Also, minimum amounts are also on the https://chainflip-broker.io/assets endpoint as `minimalAmountNative`

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/constants.ts:
```diff
@@ -0,0 +1,69 @@
+export const CHAINFLIP_DCA_SWAP_SOURCE: SwapSource = `${SwapperName.Chainflip} • DCA`
+export const CHAINFLIP_DCA_BOOST_SWAP_SOURCE: SwapSource = `${SwapperName.Chainflip} • DCA • Boost`
+
+export const usdcAsset: Asset = {
```
That's absolutely fine @CumpsD! 
Just to disambiguate here, what's the two actual flows when fees will be denominated in USDC? i.e i can see the condition is `fee.type === 'network'`and `fee.type === 'liquidity' && fee.asset === 'usdc.eth'` but not sure what this means exactly.

Asking because we may be able to simply use `buyAsset` / `sellAsset` if one of these is USDC

CumpsD replied:
Network fee is always USDC, it is the 0.10% buy & burn fee the chainflip network uses.

Liquidity is the LP fee, which I would have to double check but I believe the second leg is always USDC. It's possible this `if (fee.type === 'liquidity' && fee.asset === buyChainflipChainKey) return buyAsset` might actually never be hit.

Will double check!

CumpsD replied:
Yep, so the second leg seems to be always usdc.eth for the liquidity fee.

gomesalexandre replied:
Happy to keep the second branch for the sake of paranoia. Tyvm for disambiguating this! Looks like indeed, we need usdc as a static asset then, which means keeping it here as a const is absolutely fine.

CumpsD replied:
> Happy to keep the second branch for the sake of paranoia. Tyvm for disambiguating this! Looks like indeed, we need usdc as a static asset then, which means keeping it here as a const is absolutely fine.

Pretty much the reason I added all branches too :D

CumpsD commented on packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -164,20 +165,22 @@ export const getTradeQuote = async (
 
-      case CHAIN_NAMESPACE.Solana: {
+      // TODO(gomes): Cosmos SDK too
+      case CHAIN_NAMESPACE.CosmosSdk:
```
Will this ever be the case? No idea what `CosmosSdk` is, but chainflip doesnt support any cosmos chains (right now)

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:
```diff
@@ -0,0 +1,188 @@
+      gasPrice: unsignedTxInput.gasPrice,
+    }
+  },
+  getUnsignedUtxoTransaction,
```
This should come back to this module as well since we'll need to add to trade quote meta for status polling

CumpsD commented on packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:
```diff
@@ -1,28 +1,193 @@
+
+    // Assume any state that isn't completed is pending.
+    // Going forward, we will want to add granular status polling similar to what we do for THOR.
+    if (state !== 'completed' || !swapEgress?.transactionReference) {
```
`completed` or `failed` imho

gomesalexandre replied:
Now irrelevant since we use egress transactionReference as heuristics 🙏🏽

0xean commented:
@gomesalexandre - just wanted to check in on this re: getting it merged in behind a flag as we discussed last week.  I know you are out with devcon, but let us know your plans on this guy

gomesalexandre commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -203,14 +204,13 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
   }
 
-  async buildSendTransaction(input: BuildSendTxInput<KnownChainIds.SolanaMainnet>): Promise<{
+  async buildSendApiTransaction(input: BuildSendApiTxInput<KnownChainIds.SolanaMainnet>): Promise<{
```
Required for Solana in swapper - we don't have `wallet` to pass around there, and this method is what we use for all other chain families to build a Tx input without a wallet, just looks like it was ommitted for Solana?

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:
```diff
@@ -0,0 +1,343 @@
+        `&destinationAsset=${buyChainflipChainKey}` +
+        `&destinationAddress=${tradeQuote.receiveAddress}` +
+        `&boostFee=10` +
+        // TODO: Calculate minprice based on tradeQuote.slippageTolerancePercentageDecimal, step.sellAmountIncludingProtocolFeesCryptoBaseUnit, step.buyAmountAfterFeesCryptoBaseUnit
+        // `&minimumPrice=` +
+        // `&refundAddress=${from}` +
+        // `&retryDurationInBlocks=10` +
```
@CumpsD is this still a valid TODO ? Does this parameter act as a limit to ensure users aren't rugged and losing funds?

CumpsD replied:
@gomesalexandre yes, still valid! This is slippage protection, which indeed ensures users arent rugged and paying way too much on a swap.

This should be the calculated using the value from the UI:
![image](https://github.com/user-attachments/assets/e63df685-26c5-4f1f-9ca3-7710c6499b59)

CumpsD replied:
From the docs:

> A minimum accepted price is specified as the ratio between the human readable destination amount and the human readable source amount e.g. 3500 for a swap from ETH to USDC if you want to receive at least 3500 USDC per ETH. Prices are checked on the AMM level and do not include deposit or broadcast fees. To calculate a minPrice based on a [quote](https://docs.chainflip.io/swapping/integrations/javascript-sdk/swap-assets/get-quote), the estimatedPrice value is multiplied with the desired slippage tolerance. For example, quote.estimatedPrice * 99 / 100 specifies a 1% slippage tolerance relative to the quote.

See more details and examples here: https://docs.chainflip.io/swapping/integrations/javascript-sdk/swap-assets/request-deposit-address/v1#fillorkillparams-slippage-protection

gomesalexandre replied:
Tyvm @CumpsD! Done in 5da50735d021e0383b2168e0233bf60753116271, added `minimumPrice` in alongside required `refundAddress` and `retryDurationInBlocks` as refund params. 

<img width="596" alt="Screenshot 2024-11-12 at 18 46 18" src="https://github.com/user-attachments/assets/d029dacb-4969-40e6-999a-b73a65e29f72">
<img width="244" alt="Screenshot 2024-11-12 at 18 46 41" src="https://github.com/user-attachments/assets/a629f7ef-a1cb-4382-8538-be7d4bfd0d1b">
<img width="596" alt="image" src="https://github.com/user-attachments/assets/603a70ca-aacf-425d-877f-e53e9f52657c">

<img width="587" alt="image" src="https://github.com/user-attachments/assets/8e30f164-5f8e-455c-91ef-2e614077114e">
<img width="208" alt="image" src="https://github.com/user-attachments/assets/e1302e27-6975-4003-b3ec-b3d8ab9acd28">
<img width="591" alt="image" src="https://github.com/user-attachments/assets/5b1f6020-6c2d-4a45-aea9-4decc0024836">

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:
```diff
@@ -0,0 +1,343 @@
+      //       .times(new BigNumber(100).minus(tolerance).dividedBy(100))
+      //       .toFixed(assetConstants[destAsset].decimals);
+
+      // TODO: For DCA swaps we need to add the numberOfChunks/chunkIntervalBlocks parameters
```
Not for this PR, streaming swaps have a lot of ties to THORChain and will require a separate unit of work to decouple from it

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/models/ChainflipBaasQuoteQuote.ts:
```diff
@@ -0,0 +1,193 @@
+// @ts-nocheck
```
openapi codegen

CumpsD commented on packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:
```diff
@@ -0,0 +1,347 @@
+    const apiKey = config.REACT_APP_CHAINFLIP_API_KEY
+
+    const step = tradeQuote.steps[0]
+    const sellChainflipChainKey = `${step.sellAsset.symbol.toLowerCase()}.${
```
No idea if you also want this to be in an `assetIdToChainflipAsset` map? It works for the current tokens with just the symbol for now :)

gomesalexandre replied:
Yep totally fine given the low amount of supported assets IMO

gomesalexandre replied:
akschually: d097f8c50149b80579acff700bc7f8e2afbb2a41

CumpsD commented:
I went over it from my end, don't count it as an official shapeshift review :) Left some remarks (biggest one left is slippage protection), but it looks good to me from chainflip specifics!

NeOMakinG commented on packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,372 @@
+  const apiKey = deps.config.REACT_APP_CHAINFLIP_API_KEY
+
+  // Subtract the BaaS fee to end up at the final displayed commissionBps
+  let serviceCommission = parseInt(commissionBps) - CHAINFLIP_BAAS_COMMISSION
```
q: wouldn't it be better to use `bnOrZero` over `parseInt`?

CumpsD replied:
Isn't commissionBps always just a small integer number? (0-90?)

gomesalexandre replied:
Think `parseInt` is fine @NeOMakinG as `commissionBps` is not a bignumber but rather a smolnumber

NeOMakinG commented on packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,372 @@
+          getDefaultSlippageDecimalPercentageForSwapper(SwapperName.Chainflip),
+        steps: [
+          {
+            buyAmountBeforeFeesCryptoBaseUnit: singleQuoteResponse.boostQuote.egressAmountNative!,
+            buyAmountAfterFeesCryptoBaseUnit: singleQuoteResponse.boostQuote.egressAmountNative!,
+            sellAmountIncludingProtocolFeesCryptoBaseUnit:
+              singleQuoteResponse.boostQuote.ingressAmountNative!,
```
q: buy amount and sell amount is the same, doesn't it sound weird?

CumpsD replied:
I guess you can calculate by using egressAmountNative and substracting includedFees from it (filtering on types which make sense for shapeshift to count as fees)

An example repsonse:

```json
   "egressAsset": "usdc.arb",
    "egressAmount": 7729.312865,
    "egressAmountNative": "7729312865",
    "includedFees": [
      {
        "type": "broker",
        "asset": "usdc.eth",
        "amount": 11.624547,
        "amountNative": "11624547"
      },
      {
        "type": "ingress",
        "asset": "flip.eth",
        "amount": 5.897594859029325662,
        "amountNative": "5897594859029325662"
      },
      {
        "type": "network",
        "asset": "usdc.eth",
        "amount": 7.757456,
        "amountNative": "7757456"
      },
      {
        "type": "egress",
        "asset": "usdc.arb",
        "amount": 0.075728,
        "amountNative": "75728"
      },
      {
        "type": "liquidity",
        "asset": "flip.eth",
        "amount": 2.497051202570485337,
        "amountNative": "2497051202570485337"
      },
      {
        "type": "liquidity",
        "asset": "usdc.eth",
        "amount": 3.869036,
        "amountNative": "3869036"
      }
    ],
```

gomesalexandre replied:
@NeOMakinG not sure what we mean here, sellAmount is using ingress (inbound) while buyAmounts are using egress (outbounds)? 

If we mean the *after fees* bit, then that's actually correct, since those fees are paid in different assets and don't require balance (a la CoW) and are effectively a downside, as can be seen on `requiresBalance: false` in 

https://github.com/shapeshift/web/blob/5da50735d021e0383b2168e0233bf60753116271/packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts#L219
<img width="488" alt="image" src="https://github.com/user-attachments/assets/58d4a289-c823-44f0-a774-62a39703ff1f">

NeOMakinG commented on src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx:
```diff
@@ -407,8 +413,33 @@ export const useTradeExecution = (
           return
         }
-        case CHAIN_NAMESPACE.Solana:
+        case CHAIN_NAMESPACE.Solana: {
+          const adapter = assertGetSolanaChainAdapter(stepSellAssetChainId)
+          const from = await adapter.getAddress({ accountNumber, wallet })
+          const output = await execution.execSolanaTransaction({
+            swapperName,
+            tradeQuote,
+            stepIndex: hopIndex,
+            slippageTolerancePercentageDecimal,
+            from,
+            signAndBroadcastTransaction: async (txToSign: SolanaSignTx) => {
+              const signedTx = await adapter.signTransaction({
+                txToSign,
+                wallet,
+              })
+              const output = await adapter.broadcastTransaction({
+                senderAddress: from,
+                receiverAddress: tradeQuote.receiveAddress,
+                hex: signedTx,
+              })
+
+              trackMixpanelEventOnExecute()
+              return output
+            },
+          })
+          cancelPollingRef.current = output?.cancelPolling
```
Thanks for this, I'll need all those Solana bits for Jup, I started but I'll rebase on this branch to retrieve your work 💜

NeOMakinG commented on src/pages/RFOX/components/Stake/Bridge/hooks/useRfoxBridge.ts:
```diff
@@ -171,6 +172,7 @@ export const useRfoxBridge = ({ confirmedQuote }: UseRfoxBridgeProps): UseRfoxBr
       assertGetEvmChainAdapter,
       assertGetUtxoChainAdapter,
       assertGetCosmosSdkChainAdapter,
+      assertGetSolanaChainAdapter,
```
Derp...

NeOMakinG commented on src/components/TradeAssetSearch/TradeAssetSearch.tsx:
```diff
@@ -162,12 +162,10 @@ export const TradeAssetSearch: FC<TradeAssetSearchProps> = ({
     })()
 
-    const sortedChainIds = sortChainIdsByDisplayName(unsortedChainIds).filter(chainId =>
-      isSwapper ? chainId !== KnownChainIds.SolanaMainnet : true,
```
q: did we remove this because we have the security at asset filtering step?

Suggesting we disable the Solana guard on line 149 and 109 if the feature flag is on, wdyt?

gomesalexandre replied:
Yeah as discussed this was a monkey-patch to hide Solana assets from all search bits, which we should've reverted or at the very least tied to Solana flag (which is now on in all envs)

gomesalexandre replied:
38492b010512e005b54d95c207d3c6e6fef2a00

NeOMakinG reviewed (commented):
Stunning work, good job guys! 🐐 

Waiting for my comments to be discussed before doing a full runtime test as the code might change depending on what we decide to accept as review comments

By curiosity, did look a Solana swap 👀 

![Uploading image.png…]()

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:
```diff
@@ -0,0 +1,376 @@
+  }: GetUnsignedUtxoTransactionArgs): Promise<BTCSignTx> => {
+    if (!isExecutableTradeQuote(tradeQuote)) throw Error('Unable to execute trade')
+
+    const brokerUrl = config.REACT_APP_CHAINFLIP_API_URL
+    const apiKey = config.REACT_APP_CHAINFLIP_API_KEY
```
Can't do that @NeOMakinG since this is relying on `config` which is *injected* for each handler, but not available in outer scope (this is swapper, not web so can't `getConfig()` here)

NeOMakinG replied:
> Can't do that @NeOMakinG since this is relying on `config` which is _injected_ for each handler, but not available in outer scope (this is swapper, not web so can't `getConfig()` here)

Ah shit, I didn't notice it was from an injected dependency...

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:
```diff
@@ -0,0 +1,376 @@
+
+    if (!isExecutableTradeStep(step)) throw Error('Unable to execute step')
+
+    const sellChainflipChainKey = `${step.sellAsset.symbol.toLowerCase()}.${
+      chainIdToChainflipNetwork[step.sellAsset.chainId as KnownChainIds]
+    }`
+    const buyChainflipChainKey = `${step.buyAsset.symbol.toLowerCase()}.${
+      chainIdToChainflipNetwork[step.buyAsset.chainId as KnownChainIds]
+    }`
```
As discussed previously in https://github.com/shapeshift/web/pull/8049#discussion_r1837434500 and https://github.com/shapeshift/web/pull/8049#discussion_r1837990324, think that's fine for the time being. Mapping doesn't really make sense as it would need to be updated either way, our best bet is codegen for those assets a la THOR gen, but don't think it makes sense for the time being given the low amount of supported assets?

gomesalexandre replied:
gm d097f8c50149b80579acff700bc7f8e2afbb2a41

gomesalexandre commented on packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:
```diff
@@ -0,0 +1,376 @@
+        `&minimumPrice=${minimumPrice}` +
+        `&refundAddress=${sendAddress}` +
+        `&retryDurationInBlocks=10` +
+        `&commissionBps=${serviceCommission}`,
```
gm [src/components/TradeAssetSearch/TradeAssetSearch.tsx](https://github.com/shapeshift/web/pull/8049/files/5da50735d021e0383b2168e0233bf60753116271#diff-ecbcf8abd534ab358976df403cfe101ef78ee7bdc3215e1389735c6cea154062)

gomesalexandre replied:
gm https://github.com/shapeshift/web/pull/8049/commits/bcce88c16a25e6f386fdc71f9242697106d86644

NeOMakinG reviewed (changes requested):
## ETH on arb => Solana
https://jam.dev/c/a8468f78-95f9-4b56-827f-3c02517c7e94

## SOL => USDC
https://jam.dev/c/c94b692c-f086-42b5-975f-c5b5fd97928d

## USDC on Solana => USDC on ETH
`TokenOwnerOffCurveError`
https://jam.dev/c/1bb836e5-b623-4e3d-bc31-3c8d935fa079

## USDC on Solana => USDC on ARB
TokenOwnerOffCurveError
https://jam.dev/c/5fad5b65-6f39-4668-9303-01193a5df769

## USDC on ETH => BTC
https://jam.dev/c/e00f1726-156d-4f40-865a-59291ecb7ad9
<img width="508" alt="image" src="https://github.com/user-attachments/assets/97d30508-d14e-491c-a4c0-856c3e6f1c74">

## ETH => BTC
https://jam.dev/c/a93baaa6-b76d-4491-b4e7-37011acfb324

## BTC =>  ETH on arb
https://jam.dev/c/b8f389c3-1929-432e-a3e3-e7411765c40c

Weird ❌ saying "Wallet is not defined"

gomesalexandre commented:
Tyvm for the review @NeOMakinG!

re: BTC sells, as mentioned in checkboxes above, those are currently blocked until we support Taproot and this PR is only getting in for the sake of progression as previously discussed, it is by no means complete.
I've added this to testing steps for clarity as well!

re: USDC.SOL sends, nice catch - this was an issue on all envs we didn't spot just yet since we didn't try to send to PDAs before. Captured in https://github.com/shapeshift/web/issues/8124, fixed in https://github.com/shapeshift/web/pull/8125, and cherry-picked in this PR in ef0066577e05bda0b438ad4ba4b75bfe9dfe78


<img width="570" alt="image" src="https://github.com/user-attachments/assets/320980a9-551d-44c6-81ca-d215371bb67d">

NeOMakinG reviewed (approved):
I forgot for taproot! Sorry

As long as it's behind a feature flag, we can now get this in, USDC on Solana to ARB has been successful now!

https://jam.dev/c/764e361e-8c9e-467b-bdc6-69714fccf62e
